current/en-us/3. fundamentals/2. creating-components.md

```diff
@@ -89,18 +89,31 @@ export class CustomerDetail {
 
 ## The Component Lifecycle
 
-All components have a well-defined lifecycle. Below is a list of methods you can implement on your view-model in order to hook into the component lifecycle:
+Every component instance has a lifecycle that you can tap into. This makes it easy for you to perform various actions at particular times. For example, you may want to execute some code as soon as your component properties are bound, but before the component is first rendered. Or, you may want to run some code to manipulate the DOM as soon as possible after your element is attached to the document. Below is a summary of the various lifecycle callbacks that you can hook into in your component.
+
+| Lifecycle Callback | Description |
+| :--- | :--- |
+| `constructor` | When the framework instantiates a component, it calls your class's constructor, just like any JavaScript class. This is the best place to put basic initialization code that is not dependent on bindable properties. |
+| `created` | The "created" hook runs just after the constructor and can be treated very similarly. The only difference is that the component's `Controller` has been instantiated and is accessible through the `$controller` property, for advanced scenarios. |
+| `bind`    | If your component has a method named "bind", then the framework will invoke it when it has begun binding values to your bindable properties. In terms of the component hierarchy, the bind hooks execute top-down, from parent to child, so your bindables will have their values set by the owning components, but the bindings in your view are not yet set. This is a good place to perform any work or make changes to anything that your view would depend on because data still flows down synchronously. This is the best time to do anything that might affect children as well. |
+| `attached` | If your component has a method named "attached", then the framework will invoke it when it has fully attached your HTML element to the document, along with its children. In terms of the component hierarchy, the attached hooks execute top-down. |
```

Comment:
These execute bottom up.

Reply:
My lifecycle demo
https://codesandbox.io/embed/aurelia-lifecycle-demo-yy1tc?expanddevtools=1&autoresize=1&fontsize=18&hidenavigation=1&module=%2Fsrc%2Fapp.html&view=preview
shows
* constructor = top-down
* created = bottom-up
* bind = top-down
* attached = top-down

Reply:
Hmm. For some reason I thought attached executed bottom up. Of course, I could be getting confused now that we have top-down and bottom-up lifecycles in v2. Thanks for verifying!

Reply:
We have been hurt by vcurrrent top-down attached.
In parent attached callback, we use setTimeout or taskQueue to ensure to run code AFTER all children DOM is attached.
It should be the other way around. I think Aurelia 2 fixed this design. Correct me if I am wrong, in Aurelia 2, attached is now buttom-up, attaching is top-down.

Reply:
That's correct @3cp .

Reply:
@baerrach please add some explanation to this wrong design, plus some info on the altered design in upcoming Aurelia 2. We kept the wrong design in Aurelia vcurrent to avoid breaking change.

Reply:
@3cp added known issues section with your recommendation